File: backend/gemini_service.py
import os
import json
import urllib.request

class GeminiService:
    # Read from os.environ, which is populated by start.js
    GEMINI_API_KEY = os.environ.get('GEMINI_API_KEY')
    GEMINI_API_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent"

    @staticmethod
    def create_script_prompt(questions, exam_name):
        # Derive a short exam name for the call to action
        exam_name_short = "".join(word[0] for word in exam_name.split() if word.isalnum()).upper()
        if not exam_name_short:
            exam_name_short = "EXAM" # Fallback

        intro = f"Hello everyone! Today, we're going to solve some challenging questions for the upcoming {exam_name} entrance exam. Let's dive in!"
        call_to_action = f"The answer and solution will appear in the last 5 seconds of this video. Till then, if you're looking for a complete guide for {exam_name}, follow and comment '{exam_name_short}' and it will be in your DMs!"

        question_sections = []
        for i, q in enumerate(questions):
            options_str = ""
            if q.get('options'):
                options_list = [f"{key}: {value}" for key, value in q['options'].items()]
                options_str = f"Here are your options: {', '.join(options_list)}."

            question_sections.append(
                f"Question {i+1}: So, the question says: {q['question_statement']}. {options_str} "
                f"You'll have a 5-second countdown to think about it."
                # Solution is not dictated here, only mentioned in the call to action
            )

        full_prompt = (
            f"You are an AI video script generator for educational content, specifically for Instagram Reels explaining mathematical problems. "
            f"Your task is to create a concise, engaging, and clear script for a voice-over. "
            f"Ensure mathematical symbols like '$x^2$', '$\int$', '$\sum$' are spoken clearly and naturally (e.g., 'x squared', 'integral', 'summation'). "
            f"The script should flow naturally for a voice-over and adhere to the following structure:\n\n"
            f"1. Introduction: Start with an engaging intro similar to: '{intro}'\n"
            f"2. For each question:\n"
            f"   - Announce the question number.\n"
            f"   - Read the question statement clearly, pronouncing mathematical symbols. For example, 'x squared' for '$x^2$'.\n"
            f"   - If options are present, read them out.\n"
            f"   - Include a phrase about a 5-second countdown for the viewer to think.\n"
            f"3. Conclusion/Call to Action: End with a call to action similar to: '{call_to_action}'\n\n"
            f"Here are the questions and their solutions (for your reference, do not dictate solutions directly in the main script body):\n\n"
            f"{' '.join(question_sections)}\n\n"
            f"Make sure the tone is encouraging and educational. Keep the overall script concise for an Instagram Reel, ideally under 300 words."
        )
        return full_prompt

    @staticmethod
    def generate_script(prompt):
        if not GeminiService.GEMINI_API_KEY:
            raise ValueError("Gemini API Key not set in environment variables.")

        headers = {
            'Content-Type': 'application/json',
        }
        payload = {
            "contents": [
                {
                    "parts": [
                        {"text": prompt}
                    ]
                }
            ]
        }
        
        url_with_key = f"{GeminiService.GEMINI_API_URL}?key={GeminiService.GEMINI_API_KEY}"

        req = urllib.request.Request(url_with_key, headers=headers, data=json.dumps(payload).encode('utf-8'), method='POST')

        try:
            with urllib.request.urlopen(req) as response:
                response_data = json.loads(response.read().decode('utf-8'))
                if 'candidates' in response_data and len(response_data['candidates']) > 0:
                    for part in response_data['candidates'][0]['content']['parts']:
                        if 'text' in part:
                            return part['text']
                return "Failed to generate script: No text found in Gemini response."
        except urllib.error.HTTPError as e:
            error_message = e.read().decode('utf-8')
            print(f"Gemini API HTTP Error: {e.code} - {error_message}")
            raise Exception(f"Gemini API Error: {error_message}")
        except Exception as e:
            print(f"Gemini API Request Error: {e}")
            raise

File: src/services/geminiService.js
const GEMINI_API_KEY = 'AIzaSyDgShKEEeX9viEQ90JHAUBfwQqlu0c9rBw';
const GEMINI_API_URL = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro:generateContent';

export class GeminiService {
  static createScriptPrompt(questions, examName) {
    // Create a short exam acronym for the call to action
    const examAcronym = examName
      .split(' ')
      .filter(word => word.length > 2)
      .map(word => word[0])
      .join('')
      .toUpperCase();

    const intro = `Hello everyone! Today, we're going to solve some challenging questions for the upcoming ${examName} entrance exam. Let's dive in!`;
    
    const callToAction = `The answer and solution will appear in the last 5 seconds of this video. Till then, if you're looking for a complete guide for ${examName}, follow and comment '${examAcronym}' and it will be in your DMs!`;

    let questionSections = '';
    questions.forEach((q, index) => {
      let optionsText = '';
      if (q.options && typeof q.options === 'object') {
        const optionsList = Object.entries(q.options)
          .map(([key, value]) => `${key}: ${value}`)
          .join(', ');
        optionsText = `Here are your options: ${optionsList}.`;
      }

      questionSections += `Question ${index + 1}: So, the question says: ${q.question_statement}. ${optionsText} You'll have a 5-second countdown to think about it. `;
    });

    return `You are an AI video script generator for educational content, specifically for Instagram Reels explaining mathematical problems. 

Your task is to create a concise, engaging, and clear script for a voice-over that will be converted to speech.

CRITICAL INSTRUCTIONS for mathematical symbols:
- Convert ALL mathematical symbols to spoken words
- $x^2$ becomes "x squared"
- $x^3$ becomes "x cubed"  
- $\\sqrt{x}$ becomes "square root of x"
- $\\int$ becomes "integral"
- $\\sum$ becomes "summation"
- $\\frac{a}{b}$ becomes "a over b" or "fraction a over b"
- $\\pi$ becomes "pi"
- $\\alpha, \\beta, \\gamma$ become "alpha, beta, gamma"
- $\\sin, \\cos, \\tan$ become "sine, cosine, tangent"
- $\\log$ becomes "logarithm"
- $\\ln$ becomes "natural log"
- $\\infty$ becomes "infinity"
- $\\pm$ becomes "plus or minus"
- $\\leq, \\geq$ become "less than or equal to, greater than or equal to"

The script should follow this exact structure:

1. Introduction: "${intro}"

2. Questions: ${questionSections}

3. Call to Action: "${callToAction}"

Make the tone encouraging and educational. Keep it concise for an Instagram Reel, ideally under 250 words. Ensure all mathematical expressions are converted to natural speech patterns.

Generate ONLY the final script text, no additional formatting or explanations.`;
  }

  static async generateScript(prompt) {
    try {
      const response = await fetch(`${GEMINI_API_URL}?key=${GEMINI_API_KEY}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          contents: [{
            parts: [{ text: prompt }]
          }]
        })
      });

      if (!response.ok) {
        throw new Error(`Gemini API error: ${response.status}`);
      }

      const data = await response.json();
      
      if (data.candidates && data.candidates[0] && data.candidates[0].content) {
        return data.candidates[0].content.parts[0].text;
      }
      
      throw new Error('No content generated from Gemini API');
    } catch (error) {
      console.error('Error generating script:', error);
      throw error;
    }
  }
}